### PR TITLE
Allow coordinator to make topology changes

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

Add a minimal API for making topology changes. This API accepts a list of operations to achieve the desired the topology. This is not the user facing API. It should be built on top of this, where the user input should be translated to the list of operations. 

## Related issues

related #14159

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: topology/src/main/java/io/camunda/zeebe/topology/changes/TopologyChangeCoordinatorImpl.java [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+                return;
+              }
+              LOG.debug(
+                  "Applying the topology change has started. The resulting topology will be {}",
+                  simulatedFinalTopology);
+              future.complete(topologyWithPendingOperations);
+            });
+  }
+
+  private void simulateTopologyChange(
```
Comment (oleschoenburg) [inline topology/src/main/java/io/camunda/zeebe/topology/changes/TopologyChangeCoordinatorImpl.java:150]: :+1: Fantastic idea to simulate first with a noop applier

File: topology/src/main/java/io/camunda/zeebe/topology/changes/TopologyChangeCoordinatorImpl.java [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+                return;
+              }
+
+              if (currentClusterTopology.version() == version) {
+                future.complete(!currentClusterTopology.hasPendingChanges());
+              } else {
+                future.completeExceptionally(
+                    new UnknownStatus(
+                        String.format(
+                            "The topology has changed since the version %d. The current version is %d. The topology change would have been already completed.",
```
Comment (oleschoenburg) [inline topology/src/main/java/io/camunda/zeebe/topology/changes/TopologyChangeCoordinatorImpl.java:84]: :upside_down_face: Technically not correct, if `version < currentClusterTopology.version()` we could show a different error message

File: topology/src/test/java/io/camunda/zeebe/topology/changes/TopologyChangeCoordinatorImplTest.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+import java.util.concurrent.ExecutionException;
+import java.util.function.UnaryOperator;
+import org.junit.jupiter.api.Test;
+
+final class TopologyChangeCoordinatorImplTest {
+
+  final TestClusterTopologyManager clusterTopologyManager = new TestClusterTopologyManager();
+
+  @Test
+  void shouldFailOnInvalidChanges() {
```
Comment (oleschoenburg) [inline topology/src/test/java/io/camunda/zeebe/topology/changes/TopologyChangeCoordinatorImplTest.java:37]: :wrench: Add one more test to validate that the _second_ operation can still cause a failure, i.e. two leave operations for the same member.

**Review by lenaschoenburg — APPROVED** [review]
Looks good :+1: 

I really like your idea to first simulate the changes as a validation step.

**Comment by deepthidevaki** [comment]
bors merge

**Pull request merged**